### PR TITLE
[ALLUXIO-1942 ] Better support parquet file in Alluxio

### PR DESCRIPTION
**Pull request by peisun1115**
Better support parquet file in Alluxio: 
a) handle cache stream related exceptions properly in FileInStream 
b) Increase netty network timeout from 3s to 30s
c) Fix a bug related to root dir in FileSystemMaster

### CODE REVIEW

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 1]
```diff
@@ -1630,7 +1630,7 @@ private long loadDirectoryMetadata(AlluxioURI path, LoadMetadataOptions options)
       throws IOException, InvalidPathException, AccessControlException, FileDoesNotExistException {
     if (mInodeTree.inodePathExists(path)) {
       Inode inode = mInodeTree.getInodeByPath(path);
-      if (inode.isPersisted()) {
+      if (inode.isPersisted() || path.isRoot()) {
```
Comment (peisun1115) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: path.isRoot() is necessary because Alluxio considers root as existing by default.

**Comment by peisun1115** [comment]
@calvinjia  @apc999 

I have rebased the PR to master. Now it is ready for review.

File: core/common/src/main/resources/alluxio-default.properties [hunk 2]
```diff
@@ -148,14 +148,14 @@ alluxio.user.failed.space.request.limits=3
 alluxio.user.file.buffer.bytes=1MB
 alluxio.user.file.master.client.threads=10
 alluxio.user.file.readtype.default=CACHE_PROMOTE
-alluxio.user.file.cache.partially.read.block=false
+alluxio.user.file.cache.partially.read.block=true
```
Comment (peisun1115) [inline core/common/src/main/resources/alluxio-default.properties]: I have changed this to true by default because I think it is annoying to ask spark users to set this to true everytime. Let me know what you think. thanks

File: core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java [hunk 3]
```diff
@@ -136,4 +138,12 @@ private long requestSpace(long requestBytes) throws IOException {
     }
     return requestBytes;
   }
+
+  /**
+   * Release {@link #mBlockWorkerClient} and set {@link #mClosed} to true.
```
Comment (apc999) [inline core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java]: NIT: Releases

File: core/client/src/main/java/alluxio/client/file/FileInStream.java [hunk 4]
```diff
@@ -124,6 +126,9 @@ protected FileInStream(URIStatus status, InStreamOptions options) {
     mAlluxioStorageType = options.getAlluxioStorageType();
     mShouldCache = mAlluxioStorageType.isStore();
     mShouldCachePartiallyReadBlock = options.isCachePartiallyReadBlock();
+    if (mShouldCachePartiallyReadBlock) {
+      LOG.debug("Should cache partially read block enabled.");
```
Comment (apc999) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: NIT: `String.format("%s enabled", Constants.USER_FILE_CACHE_PARTIALLY_READ_BLOCK)`

File: core/client/src/main/java/alluxio/client/file/FileInStream.java [hunk 5]
```diff
@@ -321,17 +326,25 @@ protected boolean shouldUpdateStreams(long currentBlockId) {
...
+      } else {
+        mCurrentCacheStream.cancel();
+      }
+    } catch (IOException e) {
+      if (e.getCause() instanceof BlockDoesNotExistException) {
+        LOG.info("Block {} does not exist when being cancelled.", getCurrentBlockId());
+      } else if (e.getCause() instanceof InvalidWorkerStateException) {
+        LOG.info("Block {} has invalid worker state when being cancelled.", getCurrentBlockId());
+      } else {
+        LOG.info("Close or cancel throws IOExecption {}.", e.getMessage());
```
Comment (apc999) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: shall we escalate the exception there?

File: core/client/src/main/java/alluxio/client/file/FileInStream.java (hunk 4 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: done

File: core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java (hunk 3 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java]: done

File: core/client/src/main/java/alluxio/client/file/FileInStream.java (hunk 5 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: I changed this to LOG.warn now.

I don't think we should throw an exception here because read should continue though cache fails (e.g. timeout because of slow network, server unreachable because of alluxio worker is busy etc).  What do you think?

**Comment by peisun1115** [comment]
Jenkins, test this please

**Comment by peisun1115** [comment]
Jenkins, test this please

File: core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java [hunk 6]
```diff
@@ -81,9 +81,9 @@ public void cancel() throws IOException {
       mBlockWorkerClient.cancelBlock(mBlockId);
     } catch (AlluxioException e) {
       throw new IOException(e);
+    } finally {
```
Comment (apc999) [inline core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java]: Q: was this previously a bug?

File: core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java [hunk 7]
```diff
@@ -98,11 +98,13 @@ public void close() throws IOException {
         mBlockWorkerClient.cacheBlock(mBlockId);
       } catch (AlluxioException e) {
         throw new IOException(e);
+      } finally {
```
Comment (apc999) [inline core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java]: Q: was this previously a bug?

File: core/client/src/main/java/alluxio/client/file/FileInStream.java [hunk 8]
```diff
@@ -249,7 +254,7 @@ public long skip(long n) throws IOException {
     }
 
     long toSkip = Math.min(n, remaining());
-    seekInternal(mPos + toSkip);
+    seek(mPos + toSkip);
```
Comment (apc999) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: why do we change from `seekInternal` to `seek`?

File: core/client/src/main/java/alluxio/client/file/FileInStream.java (hunk 5 above)
Comment (apc999) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: My concern is that, I am not sure if the caller of `closeOrCancelCacheStream` making any assumption here: is it possible that previously the caller assume exceptions thrown to indicate the error, while now this assumption is no more true? Please double-check if callers of `closeOrCancelCacheStream` did make this assumption.

File: core/client/src/main/java/alluxio/client/file/FileInStream.java [hunk 9]
```diff
@@ -321,17 +326,28 @@ protected boolean shouldUpdateStreams(long currentBlockId) {
...
-      mCurrentCacheStream.cancel();
+    try {
+      if (mCurrentCacheStream.remaining() == 0) {
+        mCurrentCacheStream.close();
+      } else {
+        mCurrentCacheStream.cancel();
+      }
+    } catch (IOException e) {
+      if (e.getCause() instanceof BlockDoesNotExistException) {
+        LOG.info("Block {} does not exist when being cancelled.", getCurrentBlockId());
```
Comment (apc999) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: i think the three cases should be `LOG.warn`, while the last case shoule be `LOG.error`

Comment (apc999) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: and could you also comment here why for these three cases, it is fine to proceed. E.g., what could actually be happening to cause these problem?

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 1 above)
Comment (apc999) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: please let @gpang  verify  this change

File: core/client/src/main/java/alluxio/client/file/FileInStream.java (hunk 9 above)
Comment (apc999) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: Another dumb question, I guess the change here is optimized for concurrent Alluxio readers (with cache turned on), e.g., Parquet readers. Will this change affect the normal case?

**Comment by apc999** [comment]
I left a bunch questions. Thanks. Now I agree with Calvin that it might be worth making a subclass of FileInStream that can cache partially-read blocks. I don't think it should be done in this PR, but I will view this as a tech debt.

File: core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java (hunk 6 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java]: I think so.

File: core/client/src/main/java/alluxio/client/file/FileInStream.java (hunk 8 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: Because I think the partially read block caching should be implemented in skip as well.  Previously skip doesn't have this feature. I have not seen a problem related to skip but maybe spark use skip in the future. that will make it very hard to debug if we don't have this

File: core/client/src/main/java/alluxio/client/file/FileInStream.java (hunk 9 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: 1. If I use LOG.warn or LOG.error, the spark shell will be spammed by logs unless the user do sc.setLogLevel("ERROR").  In fact, i think this should be LOG.debug. Otherwise it might affect performance since this it happens so often in large parquet files.  
2. Documents added. 
3. This change optimizes for concurrent Alluxio readers with cache turned on (note that it affects concurrent readers with or without partially read block caching turned on). It affects normal cases. We used to throw and exception to the reader if the caching fails (can be many reasons such as network partition, unreachable due to alluxio worker is busy etc. etc. ), but now we don't.   I think it is expected to see alluxio client to behave this way, i.e. don't crash the reader if cache fails.  For example, a spark user might prefer to have his job working instead of crashing if one AlluxioWorker lost.

Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: Another point I forgot to mention:
Before this PR, we were inconsistent in handling cache stream related exceptions. We were ignoring all exceptions (only log them) in cache.write but not in cache.cancel or cache.close. I think it is better to be consistent.

**Comment by peisun1115** [comment]
@apc999  I have addressed your comments.

About subclassing FileInStream, I talked with Calvin last time. Personally I think we are abusing Inheritance because:
1.  CachePartiallyReadBlock is only one option in FileInStream. When we have multiple options that you want to subclass, that becomes problematic since we might want to have a combination of the options.  For example, now we want to have UnknownLengthFileInStream with partially read block caching enabled. Do you want to have a subclass of FileInStream and a subclass of UnknownLengthFileInStream?
2.  Quoting from effective java (Item 16: Unlike method invocation, inheritance violates encapsulation) 
3. When I make changes to FileInStream, I felt that the subclass UnknownLengthFileInStream makes it harder to make changes to FileInStream if UnknownLengthFileInStream overrides too many small functions. 

I think  this is a style of programming, we can discuss more and I can be convinced if you and clears my concerns.

File: core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java (hunk 7 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java]: i think so

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 1 above)
Comment (peisun1115) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: I will send another PR for this.

File: core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java (hunk 6 above)
Comment (apc999) [inline core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java]: ack

File: core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java (hunk 7 above)
Comment (apc999) [inline core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java]: ack

File: core/client/src/main/java/alluxio/client/file/FileInStream.java (hunk 8 above)
Comment (apc999) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: ack

File: core/client/src/main/java/alluxio/client/file/FileInStream.java [hunk 10]
```diff
@@ -321,17 +326,40 @@ protected boolean shouldUpdateStreams(long currentBlockId) {
...
-      mCurrentCacheStream.cancel();
+    try {
+      if (mCurrentCacheStream.remaining() == 0) {
+        mCurrentCacheStream.close();
+      } else {
+        mCurrentCacheStream.cancel();
+      }
+    } catch (IOException e) {
+      if (e.getCause() instanceof BlockDoesNotExistException) {
+        // This happens if two concurrent readers read trying to cache the same block. One cancelled
```
Comment (apc999) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: thanks for the detailed documentation here! it is very helpful

File: core/client/src/main/java/alluxio/client/file/FileInStream.java (hunk 9 above)
Comment (apc999) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: @peisun1115 , about the expected behavior on alluxio client, please double check on Calvin as he has most context of alluxio client-side code.

**Comment by apc999** [comment]
thanks for the update @peisun1115 . I have no more comments left, but leaving comments about consulting with @calvinjia and @gpang (it seems you have separate PR for him). I will watch the conversation then.

File: core/client/src/main/java/alluxio/client/file/FileInStream.java [hunk 11]
```diff
@@ -124,6 +126,9 @@ protected FileInStream(URIStatus status, InStreamOptions options) {
     mAlluxioStorageType = options.getAlluxioStorageType();
     mShouldCache = mAlluxioStorageType.isStore();
     mShouldCachePartiallyReadBlock = options.isCachePartiallyReadBlock();
+    if (mShouldCachePartiallyReadBlock) {
+      LOG.debug("CacheParitallyReadBlock is enabled.");
+    }
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: If we have a debug message about the stream options, perhaps use the `options.toString` method to get all the info?

Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: Done

File: core/client/src/main/java/alluxio/client/file/FileInStream.java (hunk 10 above)
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: For the log messages, could you make this a bit more user facing? ie. describe the expected state (if the block is expected to be cached or not)

File: core/client/src/main/java/alluxio/client/file/FileInStream.java [hunk 12]
```diff
@@ -353,11 +381,14 @@ private long getCurrentBlockId() {
    * Logs IO exceptions thrown in response to the worker cache request. If the exception is not an
    * expected exception, a warning will be logged with the stack trace.
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: Update javadoc

File: core/client/src/main/java/alluxio/client/file/FileInStream.java [hunk 13]
```diff
@@ -353,11 +381,14 @@ private long getCurrentBlockId() {
...
+  private void handleCacheStreamIOException(IOException e) {
     if (e.getCause() instanceof BlockAlreadyExistsException) {
-      LOG.warn(BLOCK_ID_EXISTS_SO_NOT_CACHED, getCurrentBlockId());
+      // This can happen if there are two readers trying to cache the same block. The first one
+      // created the block (either as temp block or committed block). The second sees this
+      // exception.
+      LOG.info(BLOCK_ID_EXISTS_SO_NOT_CACHED, getCurrentBlockId());
     } else {
-      LOG.warn(BLOCK_ID_NOT_CACHED, getCurrentBlockId(), e);
+      LOG.info(BLOCK_ID_NOT_CACHED, getCurrentBlockId());
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: Why is this changed to `info`?

File: core/client/src/main/java/alluxio/client/file/FileInStream.java [hunk 14]
```diff
@@ -580,9 +611,12 @@ private void readCurrentBlockToPos(long pos) throws IOException {
       return;
     }
 
-    byte[] buffer = new byte[Math.min(Constants.MB, (int) len)];
+    // Do not set the buffer size too small to avoid slowing down seek by too much.
+    byte[] buffer = new byte[Math.min(Constants.MB * 128, (int) len)];
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: Make this value configurable?

File: core/client/src/main/java/alluxio/client/file/FileInStream.java (hunk 12 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: done

File: core/client/src/main/java/alluxio/client/file/FileInStream.java (hunk 13 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: It prints some unnecessary msgs on the spark-shell. I changed this back to LOG.warn but removed the exception in it.  I forgot the exact situation this happened. I might need to change this back to info after the test. For now, it is warn

File: core/client/src/main/java/alluxio/client/file/FileInStream.java (hunk 14 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileInStream.java]: done

**Comment by peisun1115** [comment]
@calvinjia  I have addressed your comments. PTAL. Thank you

**Comment by calvinjia** [comment]
@peisun1115 Looks good overall, could you add the new option fields to `InStreamOptionsTest`?

**Comment by peisun1115** [comment]
@calvinjia  tests updated

**Comment by calvinjia** [comment]
LGTM

File: docs/_data/table/cn/user-configuration.yml [hunk 15]
```diff
@@ -27,6 +27,8 @@ alluxio.user.file.writetype.default:
   `CACHE_THROUGH` (尽量缓冲数据，同时同步写入到底层文件系统), `THROUGH` (不缓冲数据，同步写入到底层文件系统)。
 alluxio.user.file.cache.partially.read.block:
   当读取类型是 `CACHE` 或者 `CACHE_PROMOTE` 的时候, 如果设置这个变量为真, 没有完全读取的数据块也会被存到Alluxio内.
+alluxio.user.file.seek.buffer.size.bytes:
+  在文件seek操作中使用的缓存大小。
```
Comment (apc999) [inline docs/_data/table/cn/user-configuration.yml]: also indicate that this is only used when alluxio.user.file.cache.partially.read.block is enabled?

**Comment by apc999** [comment]
one minor format comment, otherwise, LGTM

File: docs/_data/table/cn/user-configuration.yml (hunk 15 above)
Comment (peisun1115) [inline docs/_data/table/cn/user-configuration.yml]: done

**Comment by peisun1115** [comment]
@apc999  Addressed

**Comment by apc999** [comment]
LGTM
